crates/test-programs/tests/wasm_tests/runtime/tokio.rs

```diff
@@ -0,0 +1,138 @@
+use anyhow::Context;
+use std::path::Path;
+use wasi_common::pipe::WritePipe;
+use wasmtime::{Config, Engine, Linker, Module, Store};
+use wasmtime_wasi::tokio::{Wasi, WasiCtxBuilder};
+
+pub fn instantiate(data: &[u8], bin_name: &str, workspace: Option<&Path>) -> anyhow::Result<()> {
+    let stdout = WritePipe::new_in_memory();
+    let stdout_ = stdout.clone();
+    let stderr = WritePipe::new_in_memory();
+    let stderr_ = stderr.clone();
+
+    let r = tokio::runtime::Runtime::new()
+        .expect("create runtime")
+        .block_on(async move {
+            let mut config = Config::new();
+            config.async_support(true);
+            config.consume_fuel(true);
+            Wasi::add_to_config(&mut config);
+            let engine = Engine::new(&config)?;
+            let store = Store::new(&engine);
+
+            // Create our wasi context.
+            // Additionally register any preopened directories if we have them.
+            let mut builder = WasiCtxBuilder::new();
+
+            builder = builder
+                .arg(bin_name)?
+                .arg(".")?
+                .stdout(Box::new(stdout_))
+                .stderr(Box::new(stderr_));
+
+            if let Some(workspace) = workspace {
+                println!("preopen: {:?}", workspace);
+                let preopen_dir = unsafe { cap_std::fs::Dir::open_ambient_dir(workspace) }?;
+                builder = builder.preopened_dir(preopen_dir, ".")?;
+            }
+
+            #[cfg(windows)]
+            {
+                builder = builder
+                    .env("ERRNO_MODE_WINDOWS", "1")?
+                    .env("NO_DANGLING_FILESYSTEM", "1")?
+                    .env("NO_FD_ALLOCATE", "1")?
+                    .env("NO_RENAME_DIR_TO_EMPTY_DIR", "1")?
+            }
+            #[cfg(all(unix, not(target_os = "macos")))]
+            {
+                builder = builder.env("ERRNO_MODE_UNIX", "1")?;
+            }
+            #[cfg(target_os = "macos")]
+            {
+                builder = builder
+                    .env("ERRNO_MODE_MACOS", "1")?
+                    .env("NO_FD_ALLOCATE", "1")?;
+            }
```

Comment:
This is my personal first time seeing this, although I see that this is simply copied over from the cap-std-sync version. Could this be deduplicated between those two versions perhaps? (also if you know what this does can you add a comment here about that? I have no idea what this is configuring...)